RFC-0005-structured-kernel-definitions.md

```diff
@@ -0,0 +1,323 @@
+      out of this dispatch key will get this logic applied to them. This
+      key is skipped for core operators, as this checking logic is fused
+      directly into the backend kernel. External backends can also opt
+      to fuse this logic in.
+
+## Details
+
+Let’s suppose you want to write a new operator from scratch. The first
+thing you will do is add a native_functions.yaml declaration for it.
+Let’s take `upsample_nearest1d` as the example for this post.  You will
+write an entry like this, describing the *functional* version of this
+operator:
+
+```
+- func: upsample_nearest1d(Tensor self, int[1] output_size, float? scales=None) -> Tensor
+  dispatch:
+    CPU: upsample_nearest1d_cpu
+    CUDA: upsample_nearest1d_cuda
+```
+
+Because you are a conscientious implementor, you also want to provide an
+out= variant of this function. This version gets a separate entry:
+
+```
+- func: upsample_nearest1d.out(Tensor self, int[1] output_size, float? scales=None, *, Tensor(a!) out) -> Tensor(a!)
+  dispatch:
+    CPU: upsample_nearest1d_out_cpu
+    CUDA: upsample_nearest1d_out_cuda
+```
+
+Ordinarily, these two declarations would cause function signatures for
+`upsample_nearest1d_cpu`, etc., to be generated into NativeFunctions.h,
+and then you would go ahead and implement them.  However, we propose a
+new *structured* format for writing kernels. We’ll do this by marking
+the out version of this operator as structured and deleting dispatch
+entries from the functional version (the functional operator is
+*implicitly* associated with the out-of-place version in the same way
+derivatives.yaml schemas are matched with their functional/inplace/out
+variants):
+
+```
+- func: upsample_nearest1d(Tensor self, int[1] output_size, float? scales=None) -> Tensor
+  # [NEW] dispatches for this function are omitted
+
+- func: upsample_nearest1d.out(Tensor self, int[1] output_size, float? scales=None, *, Tensor(a!) out) -> Tensor(a!)
+  structured: True  # [NEW]
+  dispatch:
+    CPU: upsample_nearest1d_structured_cpu
+    CUDA: upsample_nearest1d_structured_cuda
+```
+
+Structured definitions require a different set of functions to be
+written to implement the operator.  In this particular case, the
+functions you have to implement are:
+
+```
+namespace meta {
+ TensorMeta upsample_nearest1d(
+ const Tensor& self, IntArrayRef output_size, optional<double> scales);
+ }
+ 
+ namespace native {
+ // Precondition: out is an allocated and appropriately sized tensor;
+ // all shape checks have passed, device guards have been set,
+// version counter bumps are all handled, etc...
+ void upsample_nearest1d_structured_cpu(
+const TensorMeta& out_meta, const Tensor& out, const Tensor& self, IntArrayRef output_size, optional<double> scales);
+ void upsample_nearest1d_structured_cuda(
+const TensorMeta& out_meta,  const Tensor& out, const Tensor& self, IntArrayRef output_size, optional<double> scales);
```

Comment:
maybe worth noting somewhere in here that you're limiting the example to CPU and CUDA for expository purposes? the use of explicit dispatch keys above should probably imply to the careful reader that this setup is parameterized over all backend dispatch keys in the usual way (IIUC), but if so probably useful to make it explicit